src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs

```diff
@@ -15,242 +14,273 @@
-                // then we don't merge this.  We don't want to be munging different types of
-                // escape sequences in these strings, so we only support combining the string
-                // tokens if they're all the same type.
-                var firstStringToken = stringLiterals[0].GetFirstToken();
-                isVerbatimStringLiteral = syntaxFacts.IsVerbatimStringLiteral(firstStringToken);
-                if (stringLiterals.Any(
-                        static (lit, arg) => arg.isVerbatimStringLiteral != arg.syntaxFacts.IsVerbatimStringLiteral(lit.GetFirstToken()), (syntaxFacts, isVerbatimStringLiteral)))
-                {
+                if (isVerbatimStringLiteral != syntaxFacts.IsVerbatimStringLiteral(stringLiterals[i].GetFirstToken()))
                     return;
-                }
             }
-
-            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var interpolatedString = CreateInterpolatedString(document, isVerbatimStringLiteral, pieces);
-            context.RegisterRefactoring(
-                CodeAction.Create(
-                    FeaturesResources.Convert_to_interpolated_string,
-                    _ => UpdateDocumentAsync(document, root, top, interpolatedString),
-                    nameof(FeaturesResources.Convert_to_interpolated_string)),
-                top.Span);
         }
 
-        private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, SyntaxNode top, SyntaxNode interpolatedString)
-        {
-            var newRoot = root.ReplaceNode(top, interpolatedString);
-            return Task.FromResult(document.WithSyntaxRoot(newRoot));
-        }
+        var piecesArray = pieces.ToImmutable();
+        context.RegisterRefactoring(
+            CodeAction.Create(
+                FeaturesResources.Convert_to_interpolated_string,
+                cancellationToken => UpdateDocumentAsync(document, top, isVerbatimStringLiteral, piecesArray, cancellationToken),
+                nameof(FeaturesResources.Convert_to_interpolated_string)),
+            top.Span);
+    }
+
+    private async Task<Document> UpdateDocumentAsync(
+        Document document, SyntaxNode top, bool isVerbatimStringLiteral, ImmutableArray<SyntaxNode> pieces, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var interpolatedString = await CreateInterpolatedStringAsync(
+            document, isVerbatimStringLiteral, pieces, cancellationToken).ConfigureAwait(false);
+
+        return document.WithSyntaxRoot(root.ReplaceNode(top, interpolatedString));
+    }
 
-        protected SyntaxNode CreateInterpolatedString(
-            Document document, bool isVerbatimStringLiteral, ArrayBuilder<SyntaxNode> pieces)
+    protected async Task<SyntaxNode> CreateInterpolatedStringAsync(
```

Comment:
Trying to understand why this was changed to an ImmutableArray, and assumed because it was because the method was protected. So then I tried to understand why the method was protected, and only see the one caller of this, so now I'm confused in two regards.

Reply:
So, construction of the new node is now deferred to when the action is performed. So it can't be an array builder (since that gets returned to the pool). So I convert to an actual reified array so it can be captured safely.